clang/lib/AST/Expr.cpp

```diff
@@ -5318,3 +5319,13 @@ OMPIteratorExpr *OMPIteratorExpr::CreateEmpty(const ASTContext &Context,
       alignof(OMPIteratorExpr));
   return new (Mem) OMPIteratorExpr(EmptyShell(), NumIterators);
 }
+
+HLSLOutArgExpr *HLSLOutArgExpr::Create(const ASTContext &C, QualType Ty,
+                                       Expr *Base, bool IsInOut, Expr *WB,
+                                       OpaqueValueExpr *OpV) {
+  return new (C) HLSLOutArgExpr(Ty, Base, WB, OpV, IsInOut);
```

Comment:
For my education, how do you decide whether to use `C.Allocate` versus just `C`?

Reply:
Generally `C.Allocate` is used in contexts where the new operator won't know the full size you're allocating. In some AST nodes we allocate extra space after the node to store trailing pointers to sub-expressions. That can be valuable for reducing the size of the AST if a node can have a variable number of sub-expressions.

In this case we always have two sub-expressions, and we always have an `OpaqueValueExpr`. I could reduce the size of the AST node by not caching the `OpaqueValueExpr` and instead traversing down to it when we need it, but I'm unsure that memory savings is worthwhile.